File: scripts/graphs/xsymm.r

# ==================================================================================================
# This file is part of the CLBlast project. The project is licensed under Apache Version 2.0. This
# project uses a tab-size of two spaces and a max-width of 100 characters per line.
#
# Author(s):
#   Cedric Nugteren <www.cedricnugteren.nl>
#
# This file implements the performance script for the Xsymm routine
#
# ==================================================================================================

# Includes the common functions
args <- commandArgs(trailingOnly = FALSE)
thisfile <- (normalizePath(sub("--file=", "", args[grep("--file=", args)])))
source(file.path(dirname(thisfile), "common.r"))

# ==================================================================================================

# Settings
routine_name <- "xsymm"
parameters <- c("-m","-n","-layout","-side","-triangle",
                "-num_steps","-step","-runs","-precision")
precision <- 32

# Sets the names of the test-cases
test_names <- list(
  "multiples of 128",
  "multiples of 128 (+1)",
  "around m=n=512",
  "around m=n=2048",
  "layouts and side/triangle (m=n=1024)",
  "powers of 2"
)

# Defines the test-cases
test_values <- list(
  list(c( 128,  128, 102, 111, 111, 16, 128, num_runs, precision)),
  list(c( 129,  129, 102, 111, 111, 16, 128, num_runs, precision)),
  list(c( 512,  512, 102, 111, 111, 16, 1, num_runs, precision)),
  list(c(2048, 2048, 102, 111, 111, 16, 1, num_runs, precision)),
  list(
    c(1024, 1024, 101, 111, 111, 1, 0, num_runs, precision),
    c(1024, 1024, 101, 111, 112, 1, 0, num_runs, precision),
    c(1024, 1024, 101, 112, 111, 1, 0, num_runs, precision),
    c(1024, 1024, 101, 112, 112, 1, 0, num_runs, precision),
    c(1024, 1024, 102, 111, 111, 1, 0, num_runs, precision),
    c(1024, 1024, 102, 111, 112, 1, 0, num_runs, precision),
    c(1024, 1024, 102, 112, 111, 1, 0, num_runs, precision),
    c(1024, 1024, 102, 112, 112, 1, 0, num_runs, precision)
  ),
  list(
    c(   8,    8, 102, 111, 111, 1, 0, num_runs, precision),
    c(  16,   16, 102, 111, 111, 1, 0, num_runs, precision),
    c(  32,   32, 102, 111, 111, 1, 0, num_runs, precision),
    c(  64,   64, 102, 111, 111, 1, 0, num_runs, precision),
    c( 128,  128, 102, 111, 111, 1, 0, num_runs, precision),
    c( 256,  256, 102, 111, 111, 1, 0, num_runs, precision),
    c( 512,  512, 102, 111, 111, 1, 0, num_runs, precision),
    c(1024, 1024, 102, 111, 111, 1, 0, num_runs, precision),
    c(2048, 2048, 102, 111, 111, 1, 0, num_runs, precision),
    c(4096, 4096, 102, 111, 111, 1, 0, num_runs, precision),
    c(8192, 8192, 102, 111, 111, 1, 0, num_runs, precision)
  )
)

# Defines the x-labels corresponding to the test-cases
test_xlabels <- list(
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "matrix sizes (m=n)",
  "layout (row/col), side (l/r), triangle (up/lo)",
  "matrix sizes (m=n)"
)

# Defines the x-axis of the test-cases
test_xaxis <- list(
  c("m", ""),
  c("m", ""),
  c("m", ""),
  c("m", ""),
  list(1:8, c("row,l,up", "row,r,up", "row,l,lo", "row,r,lo",
              "col,l,up", "col,r,up", "col,l,lo", "col,r,lo")),
  c("m", "x")
)

# ==================================================================================================

# Start the script
main(routine_name=routine_name, precision=precision, test_names=test_names, test_values=test_values,
     test_xlabels=test_xlabels, test_xaxis=test_xaxis, metric_gflops=TRUE)

# ==================================================================================================

File: src/routines/level3/xsymm.cpp

// =================================================================================================
// This file is part of the CLBlast project. The project is licensed under Apache Version 2.0. This
// project loosely follows the Google C++ styleguide and uses a tab-size of two spaces and a max-
// width of 100 characters per line.
//
// Author(s):
//   Cedric Nugteren <www.cedricnugteren.nl>
//
// This file implements the Xsymm class (see the header for information about the class).
//
// =================================================================================================

#include "routines/level3/xsymm.hpp"

#include <string>
#include <vector>

namespace clblast {
// =================================================================================================

// Constructor: forwards to base class constructor
template <typename T>
Xsymm<T>::Xsymm(Queue &queue, EventPointer event, const std::string &name):
    Xgemm<T>(queue, event, name) {
}

// =================================================================================================

// The main routine
template <typename T>
void Xsymm<T>::DoSymm(const Layout layout, const Side side, const Triangle triangle,
                      const size_t m, const size_t n,
                      const T alpha,
                      const Buffer<T> &a_buffer, const size_t a_offset, const size_t a_ld,
                      const Buffer<T> &b_buffer, const size_t b_offset, const size_t b_ld,
                      const T beta,
                      const Buffer<T> &c_buffer, const size_t c_offset, const size_t c_ld) {

  // Makes sure all dimensions are larger than zero
  if ((m == 0) || (n == 0) ) { throw BLASError(StatusCode::kInvalidDimension); }

  // Computes the k dimension. This is based on whether or not the symmetric matrix is A (on the
  // left) or B (on the right) in the Xgemm routine.
  auto k = (side == Side::kLeft) ? m : n;

  // Checks for validity of the squared A matrix
  TestMatrixA(k, k, a_buffer, a_offset, a_ld);

  // Determines which kernel to run based on the layout (the Xgemm kernel assumes column-major as
  // default) and on whether we are dealing with an upper or lower triangle of the symmetric matrix
  bool is_upper = ((triangle == Triangle::kUpper && layout != Layout::kRowMajor) ||
                   (triangle == Triangle::kLower && layout == Layout::kRowMajor));
  auto kernel_name = (is_upper) ? "SymmUpperToSquared" : "SymmLowerToSquared";

  // Temporary buffer for a copy of the symmetric matrix
  auto temp_symm = Buffer<T>(context_, k*k);

  // Creates a general matrix from the symmetric matrix to be able to run the regular Xgemm
  // routine afterwards
  auto kernel = Kernel(program_, kernel_name);

  // Sets the arguments for the symmetric-to-squared kernel
  kernel.SetArgument(0, static_cast<int>(k));
  kernel.SetArgument(1, static_cast<int>(a_ld));
  kernel.SetArgument(2, static_cast<int>(a_offset));
  kernel.SetArgument(3, a_buffer());
  kernel.SetArgument(4, static_cast<int>(k));
  kernel.SetArgument(5, static_cast<int>(k));
  kernel.SetArgument(6, static_cast<int>(0));
  kernel.SetArgument(7, temp_symm());

  // Uses the common padding kernel's thread configuration. This is allowed, since the
  // symmetric-to-squared kernel uses the same parameters.
  auto global = std::vector<size_t>{Ceil(CeilDiv(k, db_["PAD_WPTX"]), db_["PAD_DIMX"]),
                                    Ceil(CeilDiv(k, db_["PAD_WPTY"]), db_["PAD_DIMY"])};
  auto local = std::vector<size_t>{db_["PAD_DIMX"], db_["PAD_DIMY"]};
  auto kernelEvent = Event();
  RunKernel(kernel, queue_, device_, global, local, kernelEvent.pointer());

  // Synchronize now: 'DoGemm' does not accept a list of events to wait for
  kernelEvent.WaitForCompletion();

  // Runs the regular Xgemm code with either "C := AB+C" or ...
  if (side == Side::kLeft) {
    DoGemm(layout, Transpose::kNo, Transpose::kNo,
           m, n, k,
           alpha,
           temp_symm, 0, k,
           b_buffer, b_offset, b_ld,
           beta,
           c_buffer, c_offset, c_ld);
  }

  // ... with "C := BA+C". Note that A and B are now reversed.
  else {
    try {
      DoGemm(layout, Transpose::kNo, Transpose::kNo,
             m, n, k,
             alpha,
             b_buffer, b_offset, b_ld,
             temp_symm, 0, k,
             beta,
             c_buffer, c_offset, c_ld);
    } catch (BLASError &e) {
      // A and B are now reversed, so also reverse the error codes returned from the Xgemm routine
      switch(e.status()) {
        case StatusCode::kInvalidMatrixA:      throw BLASError(StatusCode::kInvalidMatrixB, e.details());
        case StatusCode::kInvalidMatrixB:      throw BLASError(StatusCode::kInvalidMatrixA, e.details());
        case StatusCode::kInvalidLeadDimA:     throw BLASError(StatusCode::kInvalidLeadDimB, e.details());
        case StatusCode::kInvalidLeadDimB:     throw BLASError(StatusCode::kInvalidLeadDimA, e.details());
        case StatusCode::kInsufficientMemoryA: throw BLASError(StatusCode::kInsufficientMemoryB, e.details());
        case StatusCode::kInsufficientMemoryB: throw BLASError(StatusCode::kInsufficientMemoryA, e.details());
        default:                               throw;
      }
    }
  }
}

// =================================================================================================

// Compiles the templated class
template class Xsymm<half>;
template class Xsymm<float>;
template class Xsymm<double>;
template class Xsymm<float2>;
template class Xsymm<double2>;

// =================================================================================================
} // namespace clblast
